File: 23. Merge k Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(vector<ListNode*>& lists, int start, int end1) {
            if(start == end1)
                return lists[start];
            int mid = start + (end1 - start)/2;
            ListNode* l1 = mergeTwoLists(lists,start,mid);
            ListNode* l2 = mergeTwoLists(lists,mid+1,end1);
        
            ListNode* result = new ListNode(0);
    		ListNode* end = result;
    		while(l1!=NULL && l2!=NULL)
    		{
    			ListNode* tmp = new ListNode(l1->val>l2->val ? l2->val : l1->val);
    			if(l1->val <= l2->val)
    				l1 = l1 -> next;
    			else 
    				l2 = l2 -> next;
    			end -> next = tmp;
    			end = end -> next;
    		}
    		if(l1==NULL)
    		{
    			end -> next = l2;
    		}
    		if(l2==NULL)
    		{
    			end -> next = l1;
    		}
    		return result -> next;
    }
        
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0)
            return NULL;
        if(lists.size()==1)
            return lists[0];
        return mergeTwoLists(lists,0,lists.size()-1);
    }
};

File: 23. Merge k Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode result = null;
        ArrayList all = new ArrayList<Integer>();
        for(int i=0; i<lists.length; i++){
            ListNode tmp = lists[i];
            while(tmp!=null){
                all.add(tmp.val);
                tmp = tmp.next;
            }
        }
        Collections.sort(all,new Comparator<Integer>(){
            @Override
            public int compare(Integer i1, Integer i2){
                return i2 - i1;
            }
        });
        for(int i=0; i<all.size(); i++){
            //int tmpint = all.get(i);
            int tmpint = Integer.parseInt(String.valueOf(all.get(i)));
            ListNode tmp = new ListNode(tmpint);
            tmp.next = result;
            result = tmp;
        }
        return result;
    }
}

//arrayList 用 add and get
//对arraylist 和 array 排序，用collections.sort, 其中可以自定义comparator，注意只能返回int而不是bool
//int tmpint = Integer.parseInt(String.valueOf(all.get(i)));